pkg/kexec/memory_linux.go

```diff
@@ -0,0 +1,443 @@
+// Copyright 2015-2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package kexec
+
+import (
+	"debug/elf"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// Range represents a contiguous uintptr interval [Start, Start+Size).
+type Range struct {
+	// Start is the inclusive start of the range.
+	Start uintptr
+	// Size is the size of the range.
+	// Start+Size is the exclusive end of the range.
+	Size uint
+}
+
+// Overlaps returns true if r and r2 overlap.
+func (r Range) Overlaps(r2 Range) bool {
+	return r.Start < (r2.Start+uintptr(r2.Size)) && r2.Start < (r.Start+uintptr(r.Size))
+}
+
+// IsSupersetOf returns true if r2 in r.
+func (r Range) IsSupersetOf(r2 Range) bool {
+	return r.Start <= r2.Start && (r.Start+uintptr(r.Size)) >= (r2.Start+uintptr(r2.Size))
+}
+
+// Disjunct returns true if r and r2 does not overlap.
+func (r Range) Disjunct(r2 Range) bool {
+	return !r.Overlaps(r2)
+}
+
+// pool stores byte slices pointed by the pointers Segments.Buf to
+// prevent byte sclices to be collected by garbage collector.
+var pool [][]byte
+
+// Segment defines kernel memory layout.
+type Segment struct {
+	// Buf is a buffer in user space.
+	Buf Range
+	// Phys is  physical address of kernel.
+	Phys Range
+}
+
+func (s Segment) String() string {
+	return fmt.Sprintf("virt: %#x + %#x | phys: %#x + %#x", s.Buf.Start, s.Buf.Size, s.Phys.Start, s.Phys.Size)
+}
+
+func ptrToSlice(ptr uintptr, size int) []byte {
+	var data []byte
+
+	sh := (*reflect.SliceHeader)(unsafe.Pointer(&data))
+	sh.Data = ptr
+	sh.Len = size
+	sh.Cap = size
+
+	return data
+}
+
+func (s *Segment) tryMerge(s2 Segment) bool {
+	if s.Phys.Disjunct(s2.Phys) {
+		return false
+	}
+
+	// Virtual memory ranges should never overlap,
+	// concatenate ranges.
+	a := ptrToSlice(s.Buf.Start, int(s.Buf.Size))
+	b := ptrToSlice(s2.Buf.Start, int(s2.Buf.Size))
+
+	merged := append(a, b...)
+	s.Buf.Start = uintptr((unsafe.Pointer(&merged[0])))
+	s.Buf.Size = uint(len(merged))
+	pool = append(pool, merged)
```

Comment:
I'm a little confused as to why this is necessary. `s` and `s2` already have to be pointing to data that must not be garbage collected, right? so iff this pool has any effect, then the same data might also be garbage collected right before it gets added to the pool. To me that means the addition to the pool should be at the time of creation of `s` and `s2`, which makes this pool ineffectual. 

Maybe I've misunderstood how the Go garbage collector works, though.